### DISCUSSION TITLE
Produce a curve by intersecting two surfaces?

### DISCUSSION THREAD

**Discussion by in1291** [nortikin/sverchok#4868] [Q&A]
I was wondering if its currently possible to achieve the following: Generate a curve as a result of the intersection of two surfaces? 
In the following example, the yellow line shows what would be this resulting curve:

![image](https://user-images.githubusercontent.com/42136950/211391549-2cfd1cf7-1dd4-43a6-80de-de6eccf3330c.png)

**Comment by portnov** [comment]
No, unfortunately there is currently no clear way to do this, at least in core Sverchok.
You can try to find some workaround which would work in your particular situation.

If you have FreeCAD libraries installed, AND your curves are NURBS, it's probably possible to write a scripted node which would calculate your curve.

**Comment by rendetto** [comment]
@in1291 you can try with Solid Boolean and then extract the specific edge. You just have to manually search for it.

![Surface-Surface_Intersect](https://user-images.githubusercontent.com/66558924/211486975-0651fdff-ba9f-4219-8714-cb609e8deabf.jpg)

**Reply by portnov** [reply]
Huh. Yesterday I tried Solid boolean, but got nothing. But I was using Intersect mode, not Difference...

**Reply by rendetto** [reply]
BTW I found this https://wiki.freecadweb.org/Part_Boolean, and there is an option for finding all the Sections(Intersections) (see also https://wiki.freecadweb.org/Part_Section)
Found also an API method **section(TopoShape)** (https://wiki.freecadweb.org/TopoShape_API) - could this be the exact solution?

**Reply by rendetto** [reply]
Yep! That's the solution:
![PartSection](https://user-images.githubusercontent.com/66558924/211523119-71060748-682c-439e-a0e8-841017022632.jpg)

![PartSection2](https://user-images.githubusercontent.com/66558924/211527186-f40c11a9-a460-4f88-9cb2-658f6d371238.jpg)

**Reply by rendetto** [reply]
@portnov  is it ok if I try adding the option in the node? :)

**Reply by portnov** [reply]
@rendetto you mean to "Solid boolean"? not sure. Probably this deserves a separate node.
Because if you do `solid1.section(solid2)`, where solid1 is, for example, a cube, and solid2 is, for example, a sphere, you will probably get something strange, I guess (?). So probably it should be a separate node, which accepts surfaces (SvSolidFaceSurface's, to be exact), not any Solids.

**Reply by rendetto** [reply]
"Solid Boolean" fortunately accepts surfaces(and also curves), I was using this obscure feature before but never was aware how universal it is.
What I was planning to do is very simple (just adding the option "Section") and looks like that:

![Intersect_Surfaces](https://user-images.githubusercontent.com/66558924/211596543-cbd94f86-efa8-4f6b-9036-7cae40df3543.jpg)

![OptionSectionAdded](https://user-images.githubusercontent.com/66558924/211596958-d314ced7-2ea8-4d9c-8d0d-cde6536bf76d.jpg)

Intersection curves look pretty ok between Cube and a Sphere:
![IntersectSphereAndCube](https://user-images.githubusercontent.com/66558924/211599336-e4727848-f94e-467d-9353-52833e355fd0.jpg)

It works for intersecting curves:
![IntersectingCurves](https://user-images.githubusercontent.com/66558924/211601875-db9e0cac-c63f-4c9b-b39f-5fad3b8567ee.jpg)

And between surfaces and multiple curves:
![IntersectingSurfaceWithMultipleCurves](https://user-images.githubusercontent.com/66558924/211604634-26980f01-f94e-4b6e-9950-929a374f4f3b.jpg)

...Surface can intersect a solid cube without problems as well:
![IntersectingSurfaceWithSolidCube](https://user-images.githubusercontent.com/66558924/211606995-88b9b235-e61f-423a-9667-2149a1431b00.jpg)

**Reply by portnov** [reply]
Hmm, very nice.
But the output always contains only edges? Then probably output socket should be of Curve type, not Solid in this mode.

I'm just a bit confused about adding this mode in Solid Boolean. When we say Boolean, we usually mean Intersect / Union / Difference, not something else. If someone will search for this operation, they will probably not be expecting to find it among boolean operations.  So maybe separate node will look better, even if it will be partly a copy-paste of Solid Boolean.

**Comment by nortikin** [comment]
solid-soid = solid # solid boolean
surface-surface = curve # solidbooleat
curve-curve = point # curve booleat

**Reply by rendetto** [reply]
> solid-soid = solid # solid boolean 
surface-surface = curve # solidbooleat 
curve-curve = point # curve booleat

there are also:
solid-curve = **point**(common case) or **curve** (if the curve lies on the solid)
solid-surface = **curve**(common case) or **point**(when both are only touching)
curve-surface = **point**(common case) or **curve**(if the curve lies on the surface)

and also those first could produce:
solid-solid = **curve**(common case) or **point** in special cases
surface-surface = **curve**(common case) or **point** in special cases
curve-curve = **point**(common case) or **curve** if curves coincide at a certain segment

**Reply by nortikin** [reply]
> > solid-soid = solid # solid boolean
> > surface-surface = curve # solidbooleat
> > curve-curve = point # curve booleat
> 
> there are also: solid-curve = point solid-surface = curve curve-surface = point


You won, it is solid boolean

**Reply by portnov** [reply]
@rendetto okay, we are waiting for PR to include this option to Solid Boolean :)

**Reply by rendetto** [reply]
Umm... had to change the plan after some thorough tests.

@portnov  You're right about the output! "Section" will outputs only edges (and eventually points) and never surfaces (despite my theoretical conclusion). I tried some tests with overlapping surfaces and got only the contour edges of the inner surface. The tests with overlapping curves created curves, so it's predictable. Also touching objects at points also creates only vertices, so this was also correct assumption. 
I'll try making a new "Section Node" (that outputs Curve type) as a partly a copy-paste of Solid Boolean as you proposed.
Wish me luck :)

**Reply by rendetto** [reply]
So I'm in trouble. I started thinking back to just insert the option into "Solid Boolean" node and make it's documentation more detailed. There are several implications due to the mixed input/output that I faced with creating a dedicated "Section Node".

**Implication 1:** As we can mix solids, surfaces and curves, a "Solid" is not adequate name of the sockets as we can input other types of geometry - to make things more universal I propose to name the input "Shape" (as in OCC a "Shape" unites all possible types of objects). 

**Implication 2:** "Accumulate nested" is not compatible with curves as input. If we try adding a curve to the list of input shapes an OCC error is raised instead. So we can only intersect two shapes if we want to include curve/s as input. I have no solution for this problem.

**Implication 3:** As "Section" operation outputs not only curves but also points, there is no sense to limit the output only to curve objects. I think that the original output as Shape object (which contains Edges and Vertices) is more appropriate.   

Does it look a bit inconsistent compared to the other nodes from the group?
![SectionNode](https://user-images.githubusercontent.com/66558924/211915880-71fb7954-953e-48b0-ac80-c6cbba58fff6.jpg)

Note that those implications will be still present in case we decide to only add the "Section" as an option in the "Solid Boolean" node.

So which option sounds more reasonable:
1. Make the "Solid Section" node as described above? (easier/intuitive for the new users to discover)
2. Only Add the Section as an option to the "Solid Boolean" node? (this is how it's done in FreeCAD, and those users will possibly search for the option there)
3. Implement both options above? (both FreeCAD and new users can discover what they are used to)

**Reply by portnov** [reply]
> Implication 2

Well, if we can't implement "accumulate nested" correctly, let's just not support it for this operation.

> Implication 3

Maybe it will be better to have two outputs, "Curves" and "Vertices"? Or Shape output object provides more possibilities than just decomposing it to edges and vertices?

**Reply by rendetto** [reply]
Thanks @portnov! Is it ok like that?
![SectionNodeMenuV2](https://user-images.githubusercontent.com/66558924/212058287-c0426edb-7f56-425d-8e7c-cc99ff8700c7.jpg)
![SectionNodeV2](https://user-images.githubusercontent.com/66558924/212060254-eaffa665-78d0-401b-a5f5-31e472ce8db6.jpg)

**Reply by portnov** [reply]
Yeah, looks good!

**Reply by rendetto** [reply]
I managed to test and stabilize the node finally. What left is the documentation (which is almost done) but I'm a bit confused. Where or how should I upload the images I want to show in the docs?

**Reply by portnov** [reply]
Usually we just upload them to github (for example into description of PR, or into comments) and give links to them.

**Comment by rendetto** [comment]
I finished the node https://nortikin.github.io/sverchok/docs/nodes/solid/solid_section.html

There is something to consider though. We have to be aware that the resulting curves are of degree 1 - high resolution polylines.

So if smoother results are needed the intersection curves could be approximated with the existing methods.

Cheers
